### PR TITLE
Pulse 3507: Add retrying functionality to all Redis and Kafka methods

### PR DESCRIPTION
**Pull request by rdempsey**
Added retrying functionality to all functions connecting to and using Redis and Kafka. Used the tenacity library.

Successfully running in the development environment.

**Testing**

Get the code

- `git clone git@github.com:istresearch/traptor.git`
- `git checkout pulse-3507-retries`

Set up your environment

- Create and activate a virtual environment (venv or Anaconda)
- Rename `traptor.env.sample` to `traptor.env` and fill in all fields. Some defaults are provided.
- Install the requirements: pip install -r requirements.txt

Run the tests

- In the root directory: python -m pytest -vv --cache-clear --cov=traptor --cov-report html

All tests should pass.

### CODE REVIEW

File: traptor/traptor.py [hunk 1]
```diff
@@ -138,6 +139,11 @@ def _setup_birdy(self):
                                         self.apikeys['ACCESS_TOKEN_SECRET']
                                         )
 
+    @retry(wait=wait_exponential(multiplier=1, max=10),
+           stop=stop_after_attempt(3),
+           reraise=True,
+           retry=retry_if_exception_type(KafkaUnavailableError)
```
Comment (jasonrhaas) [inline traptor/traptor.py]: This only handles the `KafkaUnaviailableError` scenario right?  What about connect timeout, network drops, or other errors?  What I did in vkollect is creating a new class called TenaciousProducer that handles all the kafka errors, trying the producer if it fails, and re-creating it as a last try.

What you have here is fine but I think we need a more all encompassing retry solution for kafka.

File: traptor/traptor.py [hunk 2]
```diff
@@ -213,6 +219,11 @@ def _setup(self):
         # Create the locations_rule dict if this is a locations traptor
         self.locations_rule = {}
 
+    @retry(wait=wait_exponential(multiplier=1, max=10),
+           stop=stop_after_attempt(3),
+           reraise=True,
```
Comment (jasonrhaas) [inline traptor/traptor.py]: Is the `reraise` needed here?  My understanding is that the reraise is for if you want to handle the case when the retry fails... and do something after that.

File: traptor/traptor.py [hunk 3]
```diff
@@ -235,7 +246,7 @@ def _create_birdy_stream(self):
                 })
                 dd_monitoring.increment('twitter_error_occurred',
                                         tags=['error_type:twitter_api_error'])
-                sys.exit(3)
+                # sys.exit(3)
```
Comment (jasonrhaas) [inline traptor/traptor.py]: After it retires 3 times what happens?  Just tracebacks or?

File: traptor/traptor.py [hunk 4]
```diff
@@ -863,6 +909,33 @@ def _send_heartbeat_message(self):
...
+        :param enriched_data: the enriched data to send
+        """
+        try:
+            self.logger.info("Attempting to send tweet to kafka", extra={
+                'tweet_id': tweet.get('id_str', None)
+            })
+            future = self.kafka_conn.send(self.kafka_topic, enriched_data)
+            future.add_callback(self.kafka_success_callback, tweet)
+            future.add_errback(self.kafka_failure_callback)
+        except KafkaUnavailableError:
```
Comment (jasonrhaas) [inline traptor/traptor.py]: Does this work with the `@retry` if you catch the error like this?  It hasn't worked for me.

File: traptor/traptor.py (hunk 2 above)
Comment (rdempsey) [inline traptor/traptor.py]: I wasn't fully sure. I'll remove this for now.

File: traptor/traptor.py (hunk 3 above)
Comment (rdempsey) [inline traptor/traptor.py]: Per our conversation I'm going to uncomment these lines.

File: traptor/traptor.py (hunk 1 above)
Comment (rdempsey) [inline traptor/traptor.py]: Agreed. I'll keep the one for now and perhaps we can turn your `TenaciousProducer` into a library and use that. This is pretty preliminary.

File: traptor/traptor.py (hunk 4 above)
Comment (rdempsey) [inline traptor/traptor.py]: Changed back to plain Exception to get the logging. The retry will handle the Kafka error.

**Pull request merged**